### ISSUE TITLE
Make hocr-tools a proper module

### ISSUE THREAD

**Issue by kba** [tmbdev/hocr-tools#42]
The README [currently](https://github.com/tmbdev/hocr-tools/blob/76fe7679329776d325a1d22cf3b8aa2105141589/README.md) states:

> Each command line program is self contained; if you have Python 2.7 with the required packages installed, it should just work. (Unfortunately, that means some code duplication; we may revisit this issue in later revisions.)

I would like to revisit this issue :smile: 

The advantages of striving to make the programs self-contained is that there is no need to install the whole project to run an individual script, provided the requirements were installed by some other means (e.g. apt-get). For simple scripts like `hocr-check` this is [really neat](https://github.com/UB-Mannheim/ocr-fileformat/blob/code-cleanup/script/validate/hocr).

The disadvantages of self-contained commands are IMHO:
* Code redundancy ([`assoc`](https://github.com/tmbdev/hocr-tools/search?utf8=%E2%9C%93&q=assoc), [`get_text`](https://github.com/tmbdev/hocr-tools/search?utf8=%E2%9C%93&q=get_text) etc.). These are small functions but it's considerable boilerplate and keeping them consistent is a hassle. This also makes it hard to spot that e.g. `get_text` has not been needed for a while.
* Embedding resources in the source code, such as the [invisible font in hocr-pdf](https://github.com/tmbdev/hocr-tools/blob/4588461045c72c5e4923cc5fabc25304b0e759d6/hocr-pdf#L109), makes it hard to add changes.
* It makes it harder to keep consistent interfaces. Some commands use optparse, others parse CLI arguments themselves, some read from STDIN on no args, some show the help page on no args, some exit with an error etc. A shared `hocrlib` module could help reduce boilerplate, though a consistent use of one of `argparse` could also remedy this situation.

In summary, I would argue for an approach with a shared library, resources in the file system and require users to properly (`setup.py`) install the tools.

What do you think? 

In particular, is anyone relying on the scripts being self-contained?

**Comment by stweil** [comment]
We could also write modular code and provide a way to build self contained command files.

**Comment by zuphilip** [comment]
I agree that the advantages of a shared library, resources are larger and we can also require that people are running a setup script (maybe add more options there). Personally, I always run the `setup.py`-script to make the command known system-wide.

Would it still be possible to go the PyPI way with that? 

> We could also write modular code and provide a way to build self contained command files.

Yes, we could maybe build them also for every release as a separate zip (resp. check if we can automate that). Alternatively, it could also be an option for the setup script to build independent python commands.

**Comment by kba** [comment]
> Would it still be possible to go the PyPI way with that?

Yes, all this can be handled by `setuptools` in `setup.py`.

> We could also write modular code and provide a way to build self contained command files.

Not sure if I understand. The commands cannot be truly self-contained, they still have external dependencies, e.g. `hocr-pdf` requires reportlab to be installed. Creating special versions of the scripts at build time, e.g. "baking in" the invisible font is possible but very finicky. But we can use setuptools to create egg distributions, Windows installers and such.
